### PR TITLE
Remove unused NodeArgs in Graph::Resolve

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: Remove unused NodeArgs in Graph::Resolve

**Motivation and Context**
- Certain optimizers (for example QDQ) will add unused NodeArgs (without type) to the graph, this will cause ORT conversion failure since we expect a NodeArg has a type (the unused NodeArg will not have type since typeinference will not run on it)
- Removing unused initializers will not remove the NodeArg come with it, leave unused pieces in the file
- Move the used_node_args and node_args_to_preserve to use NodeArg* ask key to improve perf comparing to the string hashing we are using now
- Remove it as part of the Graph::Resolve
- Updated test

### CODE REVIEW

File: onnxruntime/core/graph/graph.cc [hunk 1]
```diff
@@ -3426,6 +3426,28 @@ void Graph::CleanUnusedInitializers(const std::unordered_set<std::string>* initi
...
+  for (auto it = node_args_.cbegin(), node_args_end = node_args_.cend(); it != node_args_end; /* no increment */) {
+    auto current_entry = it++;
+    const auto& node_arg_name = current_entry->first;
+    if (!node_arg_name.empty() && used_args.find(node_arg_name) == end &&
+        (initializer_names_to_preserve == nullptr ||
+         initializer_names_to_preserve->find(node_arg_name) == initializer_names_to_preserve->cend())) {
+      LOGS(logger_, INFO) << "Removing NodeArg '" << node_arg_name << "'. It is no longer used by any node.";
+      // Need to remove the NodeArg from both value_info_ and node_args_
+      value_info_.erase(current_entry->second.get());
+      it = node_args_.erase(current_entry);
```
Comment (yufenglee) [inline onnxruntime/core/graph/graph.cc:3448]: >it [](http://example.com/codeflow?start=6&length=2)

nit: the assigment is unnessasary

**Review by yufenglee — DISMISSED** [review]
:shipit:

File: onnxruntime/core/graph/graph.cc [hunk 2]
```diff
@@ -3426,6 +3426,33 @@ void Graph::CleanUnusedInitializers(const std::unordered_set<std::string>* initi
...
+  // We also need to check the Outer Scope NodeArgs
+  for (const auto& outer_scope_node_arg_name : outer_scope_node_arg_names_) {
+    ORT_IGNORE_RETURN_VALUE(used_args.insert(outer_scope_node_arg_name));
+  }
+
+  for (auto it = node_args_.cbegin(), node_args_end = node_args_.cend(); it != node_args_end; /* no increment */) {
+    auto current_entry = it++;
+    const auto& node_arg_name = current_entry->first;
+    if (!node_arg_name.empty() && used_args.find(node_arg_name) == end &&
+        (initializer_names_to_preserve == nullptr ||
```
Comment (edgchen1) [inline onnxruntime/core/graph/graph.cc:3448]: if we are matching the NodeArg names too, the name `initializer_names_to_preserve` seems less precise now

Comment (gwang-msft) [reply onnxruntime/core/graph/graph.cc]: Moved to a new std::unordered_set `node_args_to_preserve` to use NodeArg* as the key

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**